packages/jsii-dotnet-generator/src/Amazon.JSII.Generator/TypeProxyGeneratorBase.cs

```diff
@@ -0,0 +1,101 @@
+using System.Collections.Generic;
+using System.Linq;
+using Amazon.JSII.JsonModel.Spec;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SF = Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Amazon.JSII.Generator
+{
+    public abstract class TypeProxyGeneratorBase<T> : TypeGeneratorBase<T> where T : Type
+    {
+        protected TypeProxyGeneratorBase(string package, T type, ISymbolMap symbols, INamespaceSet namespaces)
+            : base(package, type, symbols, namespaces)
+        {
+        }
+
+        protected override MemberDeclarationSyntax CreateType()
+        {
+            return SF.ClassDeclaration
+            (
+                CreateAttributes(),
+                SF.TokenList(SF.Token(SyntaxKind.InternalKeyword)),
+                GetProxyTypeNameSyntax(),
+                null,
+                CreateBaseList(),
+                SF.List<TypeParameterConstraintClauseSyntax>(),
+                SF.List(CreateMembers())
+            );
+
+            SyntaxList<AttributeListSyntax> CreateAttributes()
+            {
+                var typeOfExpression = SF.TypeOfExpression(Symbols.GetNameSyntax(Type));
+                var fullyQualifiedNameLiteral = SF.Literal(Type.FullyQualifiedName);
+
+                return SF.List(new[]
+                {
+                    SF.AttributeList(SF.SeparatedList(new[]
+                    {
+                        SF.Attribute(
+                            SF.ParseName("JsiiInterfaceProxy"),
```

Comment:
If you're re-using `JsiiInterfaceProxyAttribute` for abstract classes, it needs a better name. Maybe something like `JsiiTypeProxy`?